Zend/zend_object_handlers.c

```diff
@@ -377,39 +380,67 @@ static zend_always_inline uintptr_t zend_get_property_offset(zend_class_entry *c
 	if (EXPECTED(zv != NULL)) {
 		property_info = (zend_property_info*)Z_PTR_P(zv);
 		flags = property_info->flags;
-		if (UNEXPECTED((flags & ZEND_ACC_SHADOW) != 0)) {
-			/* if it's a shadow - go to access it's private */
-			property_info = NULL;
-		} else {
-			if (EXPECTED(zend_verify_property_access(property_info, ce) != 0)) {
-				if (UNEXPECTED(!(flags & ZEND_ACC_CHANGED))
-					|| UNEXPECTED((flags & ZEND_ACC_PRIVATE))) {
-					if (UNEXPECTED((flags & ZEND_ACC_STATIC) != 0)) {
-						if (!silent) {
-							zend_error(E_NOTICE, "Accessing static property %s::$%s as non static", ZSTR_VAL(ce->name), ZSTR_VAL(member));
-						}
-						return ZEND_DYNAMIC_PROPERTY_OFFSET;
+
+		if (flags & ZEND_ACC_PUBLIC) {
+check_changed:
+			if (!(flags & ZEND_ACC_CHANGED)) {
+no_changed:
+				if (UNEXPECTED((flags & ZEND_ACC_STATIC) != 0)) {
+					if (!silent) {
+						zend_error(E_NOTICE, "Accessing static property %s::$%s as non static", ZSTR_VAL(ce->name), ZSTR_VAL(member));
 					}
-					goto exit;
+					return ZEND_DYNAMIC_PROPERTY_OFFSET;
 				}
+				goto exit;
+			}
+			goto check_scope;
+		} else if (flags & ZEND_ACC_PRIVATE) {
+			if (property_info->ce != ce) {
```

Comment:
As an optimization, it should be possible to do something like this here:

```
			if (EG(fake_scope)) {
				scope = EG(fake_scope);
			} else {
				scope = zend_get_executed_scope();
			}
			if (property_info->ce == scope) {
				goto no_changed;
			}
			if (property_info->ce != ce) {
				/* if it's a shadow - go to access it's private */
				property_info = NULL;
			} else {
				/* Try to look in the scope instead */
				property_info = ZEND_WRONG_PROPERTY_INFO;
			}
```

That is, we don't need to look up the property in the scope property info table if we already have the shadow property from the right scope.

Reply:
done